File: 2/reader/FileReader.cpp
#include <fstream>
#include <string>
#include <random>
#include <vector>
#include "FileReader.h"
#include "../generator/Matrix2DGenerator.h"
#include "../generator/MatrixDiagonalGenerator.h"
#include "../generator/MatrixLowTriangleGenerator.h"
#include "MatrixDiagonalReader.h"
#include "MatrixLowTriangleReader.h"
#include "Matrix2DReader.h"

FileReader::FileReader(FileReaderCallback *_readerCallback) : readerCallback(_readerCallback) {
    matrixReaders[DIAGONAL_MATRIX_TYPE] = new MatrixDiagonalReader();
    matrixReaders[LOW_TRIANGLE_MATRIX_TYPE] = new MatrixLowTriangleReader();
    matrixReaders[TWO_DIMENSIONS_MATRIX_TYPE] = new Matrix2DReader();
}

void FileReader::read(std::string fileName) {
    std::ifstream is(fileName);
    std::string inputType;
    is >> inputType;
    if (inputType == FileReader::RAW_INPUT_TYPE) {
        readRaw(is);
    } else if (inputType == FileReader::RANDOM_INPUT_TYPE) {
        readRandom(is);
    }
}

void FileReader::readRandom(std::ifstream &ifstream) {
    int randomKey;
    ifstream >> randomKey;
    std::mt19937 randomGenerator(randomKey);

    std::string unused;
    ifstream >> unused;

    int count;
    ifstream >> count;

    auto generator1 = Matrix2DGenerator(randomGenerator);
    auto generator2 = MatrixDiagonalGenerator(randomGenerator);
    auto generator3 = MatrixLowTriangleGenerator(randomGenerator);
    std::vector<MatrixGenerator *> matrixGenerators = {&generator1, &generator2, &generator3};
    std::uniform_int_distribution<int> distribution(0, (int) matrixGenerators.size() - 1);
    std::uniform_int_distribution<int> matrixSizeDistribution(1, MAX_RANDOM_MATRIX_SIZE);
    for (int i = 0; i < count; i++) {
        MatrixGenerator *matrixGenerator = matrixGenerators[distribution(randomGenerator)];
        readerCallback->handle(matrixGenerator->generate(matrixSizeDistribution(randomGenerator)));
    }
}

void FileReader::readRaw(std::ifstream &ifstream) {
    std::string matrixType;
    while (ifstream >> matrixType) {
        Matrix* matrix = matrixReaders[matrixType]->read(ifstream);
        readerCallback->handle(matrix);
    }
}

FileReader::~FileReader() {
    for (auto &[k, v]: matrixReaders) {
        delete v;
    }
}

File: 3/reader/FileReader.py
from .Matrix2DReader import *
from .MatrixDiagonalReader import *
from .MatrixLowDiagonalReader import *
import generator
import random
import abc


class FileReaderCallback(object):
    @abc.abstractmethod
    def handle(self, matrix):
        "Handles read matrix"
        return


class FileReader(object):
    RAW_INPUT_TYPE = "raw"
    RANDOM_INPUT_TYPE = "random"
    MAX_RANDOM_MATRTIX_SIZE = 10

    DIAGONAL_MATRIX_TYPE = "diagonal"
    LOW_TRIANGLE_MATRIX_TYPE = "low_triangle"
    TWO_DIMENSIONS_MATRIX_TYPE = "2d"

    _matrixReaders = {}

    def __init__(self, reader_callback):
        self._matrixReaders[self.DIAGONAL_MATRIX_TYPE] = MatrixDiagonalReader()
        self._matrixReaders[self.TWO_DIMENSIONS_MATRIX_TYPE] = Matrix2DReader()
        self._matrixReaders[self.LOW_TRIANGLE_MATRIX_TYPE] = MatrixLowDiagonalReader()
        self._reader_callback = reader_callback

    def read(self, filename):
        input_stream = open(filename, 'r')
        input_type = input_stream.readline().strip()
        if input_type == self.RAW_INPUT_TYPE:
            self._read_raw(input_stream)
        elif input_type == self.RANDOM_INPUT_TYPE:
            self._read_random(input_stream)
        input_stream.close()

    def _read_random(self, input_stream):
        random_key = int(input_stream.readline())
        random.seed(random_key)

        input_stream.readline()

        count = int(input_stream.readline())
        generators = [
            generator.Matrix2DGenerator(),
            generator.MatrixLowTriangleGenerator(),
            generator.MatrixDiagonalGenerator()
        ]
        for i in range(count):
            cur_generator = generators[random.randrange(0, len(generators))]
            given_matrix = cur_generator.generate(random.randrange(1, self.MAX_RANDOM_MATRTIX_SIZE + 1))
            self._reader_callback.handle(given_matrix)

    def _read_raw(self, input_stream):
        while True:
            matrix_type = input_stream.readline().strip()
            if matrix_type == "":
                break
            else:
                given_matrix = self._matrixReaders[matrix_type].read(input_stream)
                self._reader_callback.handle(given_matrix)

        pass
